go-controller/pkg/libovsdbops/model_client.go

```diff
@@ -154,6 +175,10 @@ type operationModel struct {
 	// Model will have UUID set, and it can be used in DoAfter. This only works
 	// if BulkOp is false and Model != nil.
 	DoAfter func()
+	// Mutator to use for mutations.  For already existing records this should
+	// be set to MutateOperationUpdate.  If not explicitly set,
+	// ovsdb.MutateOperationInsert is used as default.
+	InsertMutator ovsdb.Mutator
```

Comment:
What if we assumed the `MutateOperationUpdate` behavior by default? And not have this extra parameter?

Currently we are handling this explicitly for specific tables/columns maps that we actually update changing values and the only reason we are doing so is because we were initially unaware if this counter-intuitive behavior of the DB with map mutations and took us by surprise. I still can't think of when we would like the original DB behavior.

Otherwise I would suggest this extra parameter to take a different form given that is not an actual ovsdb type and pretty specific to maps. How about `PreserveExistingMapValuesOnMutate bool` defaulting it to `false`. Pretty long but at least explicit and should not be used much.

With this change, should we update the following operations to be/use instead standard CreateOrUpdate ops (hopefully not missing any): `UpdateNBGlobalSetOptions`,`UpdateLogicalRouterSetExternalIDs`, `UpdateLogicalSwitchSetExternalIDs`, `UpdateLogicalSwitchSetOtherConfig`

Reply:
> What if we assumed the `MutateOperationUpdate` behavior by default? And not have this extra parameter?
> 

I didn't measure impact but we'd be doing double the amount of work on the server side when inserting keys for the first time.  So I think it's useful to have the possibility of choosing between plain insert and update.

> Currently we are handling this explicitly for specific tables/columns maps that we actually update changing values and the only reason we are doing so is because we were initially unaware if this counter-intuitive behavior of the DB with map mutations and took us by surprise. I still can't think of when we would like the original DB behavior.
> 
> Otherwise I would suggest this extra parameter to take a different form given that is not an actual ovsdb type and pretty specific to maps. How about `PreserveExistingMapValuesOnMutate bool` defaulting it to `false`. Pretty long but at least explicit and should not be used much.
> 

I can do that if you think it's easier to understand what it does.

> With this change, should we update the following operations to be/use instead standard CreateOrUpdate ops (hopefully not missing any): `UpdateNBGlobalSetOptions`,`UpdateLogicalRouterSetExternalIDs`, `UpdateLogicalSwitchSetExternalIDs`, `UpdateLogicalSwitchSetOtherConfig`

Sure, I can update these to use mutations if we agree on a direction above.

Reply:
> I didn't measure impact but we'd be doing double the amount of work on the server side when inserting keys for the first time. So I think it's useful to have the possibility of choosing between plain insert and update.

In most cases keys would be inserted for first time in a row insert and not with a mutate of an existing row, I think. But...

When you say `amount of work on the server side` do you refer to a probable needless mutate delete op we would be sending towards the ovsdb server? If that's the case we could improve by just sending that op for keys already existing in the column instead of sending the op blindly for all updated keys. So `buildMutationsFromFields` could receive the current value for the column and only build the strictly needed mutate delete op.

Not against providing an option though.
 
> I can do that if you think it's easier to understand what it does.

It's not really easier to understand in any case because the fact that the protocol itself has a pure insert behavior but not an update (upsert ) one is not very intuitive.

Where would we want that pure insert behavior by default? If we don't want it most cases if at all, then it is probably an option difficult to understand from the ovn-k perspective anyway and I would not go as far as masquerading a model_client specific option and behavior as a libovsdb's one (via using the ovsdb.Mutator type).

It would probably be better for me if it remained a properly documented but obscure option that would not be used unless specifically needed.

Reply:
> > I didn't measure impact but we'd be doing double the amount of work on the server side when inserting keys for the first time. So I think it's useful to have the possibility of choosing between plain insert and update.
> 
> In most cases keys would be inserted for first time in a row insert and not with a mutate of an existing row, I think. But...
> 
> When you say `amount of work on the server side` do you refer to a probable needless mutate delete op we would be sending towards the ovsdb server? If that's the case we could improve by just sending that op for keys already existing in the column instead of sending the op blindly for all updated keys. So `buildMutationsFromFields` could receive the current value for the column and only build the strictly needed mutate delete op.

This sounds like the best way to go, i'll see if I can implement this.

> 
> Not against providing an option though.
> 
> > I can do that if you think it's easier to understand what it does.
> 
> It's not really easier to understand in any case because the fact that the protocol itself has a pure insert behavior but not an update (upsert ) one is not very intuitive.
> 
> Where would we want that pure insert behavior by default? If we don't want it most cases if at all, then it is probably an option difficult to understand from the ovn-k perspective anyway and I would not go as far as masquerading a model_client specific option and behavior as a libovsdb's one (via using the ovsdb.Mutator type).

For the OVN component templates instantiated per chassis by inserting a key/value pair in the chassis' variables map, in `NB.Chassis_Template_Var`.  For services, creating a service will trigger an insert of a new variable but changing the endpoint set will trigger an update.  Both are rather common operations I guess.

> 
> It would probably be better for me if it remained a properly documented but obscure option that would not be used unless specifically needed.

Reply:
> For the OVN component templates instantiated per chassis by inserting a key/value pair in the chassis' variables map, in `NB.Chassis_Template_Var`. For services, creating a service will trigger an insert of a new variable but changing the endpoint set will trigger an update. Both are rather common operations I guess.

But when you say `insert` here, it is
`insert the key/value if the key is not already there or update the value if the key is already there`
rather than
 `insert the key/value if the key is not already there but do nothing if the key is already there`
?

My argument is that we don't have a need for that second behavior, or am I missing something?

Reply:
> > For the OVN component templates instantiated per chassis by inserting a key/value pair in the chassis' variables map, in `NB.Chassis_Template_Var`. For services, creating a service will trigger an insert of a new variable but changing the endpoint set will trigger an update. Both are rather common operations I guess.
> 
> But when you say `insert` here, it is `insert the key/value if the key is not already there or update the value if the key is already there` rather than `insert the key/value if the key is not already there but do nothing if the key is already there` ?

Thinking more about it, it's the former: insert if not there or update the value.  So doing what you suggested and automatically setting the mutation op type based on current column contents is probably the best way to go.

> 
> My argument is that we don't have a need for that second behavior, or am I missing something?

No, you're right, we don't need the second behavior indeed.

Reply:
> > > For the OVN component templates instantiated per chassis by inserting a key/value pair in the chassis' variables map, in `NB.Chassis_Template_Var`. For services, creating a service will trigger an insert of a new variable but changing the endpoint set will trigger an update. Both are rather common operations I guess.
> > 
> > 
> > But when you say `insert` here, it is `insert the key/value if the key is not already there or update the value if the key is already there` rather than `insert the key/value if the key is not already there but do nothing if the key is already there` ?
> 
> Thinking more about it, it's the former: insert if not there or update the value. So doing what you suggested and automatically setting the mutation op type based on current column contents is probably the best way to go.

@jcaamano After some struggles I decided to give up on this approach for now.  It's probably possible but I got a bit lost.  For now I went with your initial suggestion of adding `PreserveExistingMapValuesOnMutate bool` (default false) to the op model.  We can improve this but I wanted to unblock the review of the PR.

> With this change, should we update the following operations to be/use instead standard CreateOrUpdate ops (hopefully not missing any): `UpdateNBGlobalSetOptions`,`UpdateLogicalRouterSetExternalIDs`, `UpdateLogicalSwitchSetExternalIDs`, `UpdateLogicalSwitchSetOtherConfig`

Let's do this as a follow up after we decide if `PreserveExistingMapValuesOnMutate bool` is good enough for now, wdyt?

Reply:
> @jcaamano After some struggles I decided to give up on this approach for now. It's probably possible but I got a bit lost. For now I went with your initial suggestion of adding `PreserveExistingMapValuesOnMutate bool` (default false) to the op model. We can improve this but I wanted to unblock the review of the PR.

Sure. It's complicated as it requires accessing a struct by offset with reflect, and who knows, probably is less performant at the end than just sending all the mutate delete and mutate insert ops anyway.

I would be interested though in knowing where you would end up using `PreserveExistingMapValuesOnMutate=true`. If its nowhere, perhaps we don't need it either.

> Let's do this as a follow up after we decide if `PreserveExistingMapValuesOnMutate bool` is good enough for now, wdyt?

That's okay. Although I think it would be as simple as removing those methods and replace the calls to the corresponding standard Update methods.

Reply:
I use it when I create new load balancer template variables.  But maybe we can remove it from there too.   Let's discuss on the new version of the PR I guess.